README.md

```diff
@@ -1 +1,115 @@
+ * `Process` is the process ID for the process control calls: `Stdout`, `Status`, and `Abort`
+ * `Status` will return an enum of `COMPLETED|RUNNING|ERROR` and, ideally, the exit code
+
+## AuthN / AuthZ
+
+A hardcoded set of usernames (no passwords) will be mapped to a set of V2 cgroups 
+
+| cgroup name | CPU | Memory | 
+| ------------| --- | ------ |
+| **good**    | x   | x      |
+| **better**  | x   | x      |
+| **best**    | x   | x      |
+
+| username | assigned cgroup | 
+| -------- | --------------- |
+| **good** | good            |
+| **better** | better        |
+| **best** | best            |
+
+The caller can log in with the `Auth()` method using one of the whitelisted usernames. Other usernames will result in an error. 
+A username may be used across multiple sessions. A successful login will return a `SessionID` to the caller which must be used for all subsequent operations. 
+
+The Session ID is just a stringified UUID. 
+
+## Client UX
+
+The client initially will just be a golang program that runs the server through some paces (just an integration test harness). But in later releases I will try to turn this in to a simple cmdline tool with basic support for each operation. Exact syntax TBD but it may look something like:
+
+```
+> client auth sandy
+xxx-yyyy
+> client -session xxx-yyy run ls -laR
+12345
+> client stdout -session xxx-yyy -process 12345
+.
+..
+README.md
+> client status -session xxx-yyy -process 12345
+COMPLETED
+Exit code: 0
+> client abort -session xxx-yyy -process 12345
+already terminated
+```
+
+Beyond the raw client code generated by the gRPC code there will not be a separate "sugar" client library. Although it would be handy for managing the stdout streams. 
+
+## Implementation tradeoffs & security considerations
+
+### Shortcuts
+
+* Long lived gRPC services can be tricky with streaming messages. I will be ignoring all the retry/re-establish connection code. 
+* All the authenticated users will have hardcoded "roles" which map to cgroups. 
+* Knowing when the stdout streaming has completed could be tricky. In order to keep `client.Recv` from blocking I may keep sending empty messages server-side. Hacky I know. 
+* I may limit cgroup control to just the cpu controller. Supporting other resources shouldn't be a large leap. 
+* "domain" cgroup types will be used. "threaded" won't be supported. 
+
+## Security
+The CA and server cert will be generated with openssl and based on a 4096 bit RSA key pair. 
+An X.509 cert will be fabricated from these key pairs.
+
+There are a number of security concerns with such a service:
+* in order to exec commands the process will need to run privileged system calls. This requires root level access. Which means a bad actor could basically do anything. We'll need to use cgroups and resource isolation to protect against this. 
+* Setting the file permissions on the cgroup v2 files will be important for limiting child processes from manipulating the cgroup settings. 
```

Comment:
I think you're on to something here w.r.t. the child trying to do things to escape the  limitations set by the parent (library).

Can you add a small section describing the lifecycle of a process and the order of operations from receiving a start request, setting up cgroups, launching the job process, etc?

Reply:
Sure thing. 

Before I make the edits perhaps we can kick around my thinking a little? 
I was hoping to set up the cgroups per role (good, better, best) and just add the process ids as they execute. As opposed to creating the cgroup per process. I would tear them all down when the server exits. Is that a bad approach?

Reply:
I would create a dedicated cgroup per process, and tear it down when the process completes.

You can still have a pre-defined good/better/best setup, but I would just make a clone for each job started.

Otherwise, if one user with a role allowing them the "best" cgroup happens to start many jobs, they will share a single cgroup and may actually be more resource constrained than another user who runs a single process in the "good" cgroup.

Reply:
hmm, good point. I'll change my approach.

Reply:
I second the request for a little more details on how the job will get spun up and torn down.